src/openrct2-ui/windows/RideConstruction.cpp

```diff
@@ -233,2815 +180,2789 @@ static void CloseRideWindowForConstruction(RideId rideId)
         window_close(w);
 }
 
-/**
- *
- *  rct2: 0x006CB481
- */
-rct_window* WindowRideConstructionOpen()
-{
-    RideId rideIndex = _currentRideIndex;
-    CloseRideWindowForConstruction(rideIndex);
-
-    auto ride = get_ride(rideIndex);
-    if (ride == nullptr)
-        return nullptr;
-
-    _stationConstructed = ride->num_stations != 0;
-    _deferClose = false;
-
-    if (ride->type == RIDE_TYPE_MAZE)
-    {
-        return context_open_window_view(WV_MAZE_CONSTRUCTION);
-    }
-
-    auto w = WindowCreate(
-        ScreenCoordsXY(0, 29), 166, 394, &window_ride_construction_events, WC_RIDE_CONSTRUCTION, WF_NO_AUTO_CLOSE);
-
-    w->widgets = window_ride_construction_widgets;
+static void RideConstructPlacedForwardGameActionCallback(const GameAction* ga, const GameActions::Result* result);
+static void RideConstructPlacedBackwardGameActionCallback(const GameAction* ga, const GameActions::Result* result);
+static void CloseConstructWindowOnCompletion(Ride* ride);
 
-    WindowInitScrollWidgets(w);
+class RideConstructionWindow final : public Window
+{
+private:
+    Ride* _currentRide;
+
+    uint8_t _currentlyShowingBrakeOrBoosterSpeed;
+    uint32_t _currentDisabledSpecialTrackPieces;
+    track_type_t _currentPossibleRideConfigurations[32];
+    uint16_t _numCurrentPossibleRideConfigurations;
+    uint16_t _numCurrentPossibleSpecialTrackPieces;
+    bool _autoOpeningShop;
+
+    TileElement _tempTrackTileElement;
+    TileElement _tempSideTrackTileElement = { 0x80, 0x8F, 128, 128, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 };
+    TileElement* _backupTileElementArrays[5];
```

Comment:
It is dangerous to ever save the tile element pointers between ticks. This is only ever used in one function and not between ticks. Have a look at refactoring so that this is not required (i.e. it becomes a local variable). Same goes for the temp elements.